app/scripts/controllers/transactions/index.js

```diff
@@ -2809,8 +2811,20 @@ export default class TransactionController extends EventEmitter {
     this.store.updateState({ lastFetchedBlockNumbers });
   }
 
-  _hasTransactionHash(hash, transactions) {
-    return Object.values(transactions).some((tx) => tx.hash === hash);
+  /**
+   * This function checks if there is any transaction in the provided transactions object
+   * that either has a different hash than the provided txHash or has a type different from 'incoming'.
+   *
+   * @param {string} txHash - The transaction hash to compare with.
+   * @param {Object} transactions - The transactions to check in.
+   *
+   * @returns {boolean} - Returns true if there is at least one transaction that either has a different hash
+   * than the provided txHash or has a type different from 'incoming'. Otherwise, it returns false.
+   */
+  _dedupUnlessNotIncoming(txHash, transactions) {
+    return Object.values(transactions).some(
+      (tx) => tx.hash !== txHash || tx.type !== TransactionType.incoming,
```

Comment:
Currently we're keeping any transaction if there exists another transaction with a different hash, which will almost always be the case.

As we're not negating the result of this function, I think the logic should be:

```
return !Object.values(transactions).some(
      (tx) => tx.hash === txHash && tx.type === TransactionType.incoming,
);
```